### ISSUE TITLE
Handle tab completion for commands and names

### ISSUE THREAD

**Issue by rom1504** [mhsjlw/flying-squid#81]

**Comment by demipixel** [comment]
Maybe let commands make their own tab competition (array of possibilities)... Maybe also merge with section? (e.g. /we <tab complete> shows possible commands or sub-sections)

**Comment by rom1504** [comment]
Yeah that would do it (add a complete:function(alreadyWritten) {return completed;} option in commands)

I think we need to experiment a bit to know what would be convenient, but that complete option would probably work. (maybe it's possible to have a default complete for the most simple commands)

**Comment by demipixel** [comment]
How do we want to do this? We could have a complete() {} next to parse and action, but... It seems like it would be difficult to parse that. Should we just hope for the best and use that?

**Comment by rom1504** [comment]
I think such a complete function would take the incomplete parameters (everything after /command) and return what to complete with, or nothing if it cannot guess.
For example it could guess the player name used in /tp <player>
Mostly need a serv.completePlayerName that would take an incomplete name and figure out the complete one.
In order to do that, the simple (but inefficient if lot of players are connected) is looping on the player list and checking whether the username starts with the provided string. 
The efficient solution would be building a https://en.wikipedia.org/wiki/Trie of the username, but we probably don't need that for now.

**Comment by rom1504** [comment]
It might interesting to see if there's something convenient in https://www.npmjs.com/browse/keyword/trie though

**Comment by rom1504** [comment]
https://www.npmjs.com/package/triecomplete seems pretty good
